Add ThumbnailModule for attachments

Marcono1234 opened the pull request:
## Purpose
Resolves #165
Supersedes  #467

## Approach
When the module detects embedded images (e.g. `!test.png!`) in the description of newly created issues or in newly created comments it reads the image using [Apache Commons Imaging](https://commons.apache.org/proper/commons-imaging/) (as described in https://github.com/mojira/arisa-kt/issues/165#issuecomment-632961412) to determine the size of the image. If the image is wider or higher than a certain maximum `|thumbnail` is added to the image reference.

The module currently only checks for attached images, but not for URL image references, as originally intended by #467. Handling URL references is slightly more complex because `|thumbnail` does not work. Additionally a malicious user might be able to trigger a DoS attack (similar to [Slowloris](https://en.wikipedia.org/wiki/Slowloris_(computer_security))), or might obtain information about the server running Arisa, such as the IP address (might not be a problem). However, it appears the majority of embedded images references attachments, so handling URL references is probably not worth the effort.

Note that I am not very familiar with Kotlin, so the code might not follow best practices. Therefore I have marked this pull request as Draft for now, would be great if @urielsalis or someone else with more experience could have a look.

## Open questions / tasks
- Should the maximum image size be configurable in the config? (probably)
Will add a commit for this then.
- Editing the description of an issue or a comment directly after it was created might cause a race condition when the user directly makes an edit afterwards and Arisa overwrites it (likelyhood for this might be increased due to delay for downloading and reading image). Could be avoided / reduced by only acting _X_ minutes after creation, but that would not work with Arisa's current JQL for selecting relevant issues when no other update for that issue occurred.
- Not tested in MCTEST yet (don't have the test setup for this yet either)

#### Checklist
- [x] Included tests
- [ ] Updated documentation in [README](https://github.com/mojira/arisa-kt/blob/master/README.md) and [wiki](https://github.com/mojira/arisa-kt/wiki)
- [ ] Tested in MCTEST-xxx

Marcono1234 commented on build.gradle.kts:
```diff
@@ -60,6 +61,9 @@ dependencies {
     implementation("com.mojang", "brigadier", "1.0.17")
+    implementation("io.ktor", "ktor-client-core", ktorVersion)
+    implementation("io.ktor", "ktor-client-cio", ktorVersion)
+    implementation("org.apache.commons", "commons-imaging", "1.0-alpha2")
```
Uses alpha version because there exists no release yet.

Marcono1234 commented on src/main/kotlin/io/github/mojira/arisa/modules/ThumbnailModule.kt:
```diff
@@ -0,0 +1,189 @@
+                }
+            }
+
+            // TODO: Not sure if this is correct; should assert be called before
+            // modifying issue?
+            assertTrue(performedUpdate).bind()
```
Could someone more familiar with Kotlin / Arrow answer that?
(`TODO` comment also causes the build failure)

chandler05 replied:
Probably because of the semicolon, see below.

urielsalis replied:
IIRC we only save the updates and execute them after all modules have ran, so description wouldnt be updated here

urielsalis replied:
apart from that, I think its fine.

chandler05 commented on src/main/kotlin/io/github/mojira/arisa/modules/ThumbnailModule.kt:
```diff
@@ -0,0 +1,189 @@
+                }
+            }
+
+            // TODO: Not sure if this is correct; should assert be called before
```
```suggestion
            // TODO: Not sure if this is correct, should assert be called before
```
Detekt getting angered over the semicolon here.

Marcono1234 replied:
Thanks. I only added that as temporary comment to clear whether usage of assert is correct there. Have removed the comment now.

urielsalis commented on build.gradle.kts:
```diff
@@ -60,6 +61,9 @@ dependencies {
     implementation("io.arrow-kt", "arrow-fx", arrowVersion)
     implementation("com.beust", "klaxon", "5.4")
     implementation("com.mojang", "brigadier", "1.0.17")
+    implementation("io.ktor", "ktor-client-core", ktorVersion)
```
Instead of adding the full ktor library just for a call, I would use HttpUrlConnection like you do in java, ktor its pretty big and they break their API all the time

urielsalis replied:
and we can modify the download function instead to get a input stream

urielsalis commented on src/main/kotlin/io/github/mojira/arisa/modules/ThumbnailModule.kt:
```diff
@@ -0,0 +1,189 @@
+    /**
+     * [InputStream] which reads at most a certain number of bytes.
+     */
+    class LimitingInputStream(private val stream: InputStream, private val maxBytes: Long) : InputStream() {
```
I would move this to a separate file

Marcono1234 commented:
Rebased onto `master`, switched to JDK's Http Client and extracted ThumbnailModule settings to the config.
Could you please review the changes again?

urielsalis commented on build.gradle.kts:
```diff
@@ -33,6 +33,7 @@ configurations {
 
 val logBackVersion = "1.2.3"
 val arrowVersion = "0.10.4"
+val ktorVersion = "1.5.4"
```
No longer needed

Marcono1234 replied:
Thanks! Have fixed it now.
